src/visualizers/FormulaGrid.ts

```diff
@@ -0,0 +1,794 @@
+the formula returns an array that cannot otherwise be converted to a color,
+Formula Grid will attempt to convert each element to a color and take the
+(overlay) sum of the colors, which is equivalent to drawing each resulting
+color into the cell in the order they appear.
+
+On the other hand, the fill formula may instead return a plain object with keys
+drawn from the list of supported shapes:
+
+rectangle
+: Behaves the same as the default operation of the fill formula when it does
+  not return an object.
+
+square
+: Draws a square with side a configurable fraction of the shorter dimension
+  of a cell, centered in the cell.
+
+ellipse
+: Draws an ellipse with axes the horizontal and vertical center lines of the
+cell.
+
+circle
+: Draws a circle with diameter a configurable fraction of the shorter
+  dimension of a cell, centered in the cell.
+
+hexagon
+: Draws a hexagon twice the width and 4/3 the height of a cell, in which the
+  current cell is inscribed. The resulting hexagons doubly cover the plane;
+  if you take the odd cells horizontally in one row and the even cells in
+  the next, those hexagons tile the plane. Note the Triangle pre-set fill
+  order only visits such cells. When the aspect ratio of a cell is 1/√3, the
+  hexagons are regular. (Needs diagram.)
+
+triangle
+: Draws a triangle the height of a cell with horizontal base twice the width
+  of a cell, centered on the current cell. On odd rows the odd triangles point
+  up and the even triangles point down; on even rows the directions are
+  reversed. The resulting triangles tile the plane. When the aspect ratio
+  of a cell is 1/√3, the triangles are equilateral.
+
+text
+: This key represents a special shape. The value of the associated formula
+  will be converted into a string (if it is an array, the entries will be
+  individually converted to strings and concatenated). This string will
+  be written centered in the cell in a font size half the height of the cell
+  if it has four or fewer characters, or a font size one quarter the height
+  of the cell otherwise. The text color will be black or white, whichever
+  has highest contrast with the cell color.
+
+mouseover
+: Another special shape key. The value of the associated formula will be
+  converted into a string and shown in a popup with the mouse pointer hovers
+  over the cell.
+
+Except for the special text shapes, the value associated to a key will be
+interpreted as a color in the same way as the default behavior, and the shape
+will be drawn in that color. If multiple keys are specified, the shapes are
+drawn successively in the order that they appear in the object returned.
+
+The fill formula may use any of the variables _x_, _y_, _r_, _c_, and _s_
+described above; their values will be supplied by the visualizer. It may
+also use any of the following additional variables:
+
+_k_ -- the 1-based serial number of the cell in the order they are filled in
```

Comment:
"1-based" again... appears a variety of places

Reply:
I think I've got them all gone now, except in a code comment.